File: templating/core/src/main/python/jsonteng/tags/tag_map.py
# Copyright 2019 VMware, Inc.
# SPDX-License-Indentifier: Apache-2.0

from collections import OrderedDict

from .at_tag import AtTag
from .for_each_tag import ForEachTag
from .exists_tag import ExistsTag
from .len_tag import LenTag
from .one_of_tag import OneOfTag


_tag_class_map = {
    AtTag.name: AtTag,
    ExistsTag.name: ExistsTag,
    ForEachTag.name: ForEachTag,
    LenTag.name: LenTag,
    OneOfTag.name: OneOfTag
}


def get_tag_map(tag_resolver):
    """
    Return a tag map of known tags.
    :param tag_resolver: Tag resolved used by tags.
    :type tag_resolver: 'TagResolver'
    :return: A map of tags.
    :rtype: 'dict'
    """
    tag_map = OrderedDict()
    for tag_name, tag in _tag_class_map.items():
        tag_map[tag_name] = tag(tag_resolver)
    return tag_map


def add_tag(tag_name, tag):
    """
    Add a tag to the tag collection.
    :param tag_name: Tag name.
    :type tag_name: 'str'
    :param tag: Tag class.
    :type tag: 'TagBase'
    :return:
    """
    _tag_class_map[tag_name] = tag


def get_tag_names():
    """
    Get all tag names.
    :return: Tag names.
    :rtype: 'list'
    """
    return _tag_class_map.keys()

File: templating/core/src/main/java/com/vmware/jsonteng/tags/TagMap.java
// Copyright 2019 VMware, Inc.
// SPDX-License-Indentifier: Apache-2.0

package com.vmware.jsonteng.tags;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.util.HashMap;
import java.util.Hashtable;
import java.util.Map;

import com.vmware.jsonteng.TagResolver;
import com.vmware.jsonteng.TemplateEngineException;

public class TagMap {
    private static final Map<String, Class<?>> tagClassMap;

    static {
        tagClassMap = new Hashtable<>();

        tagClassMap.put(AtTag.name, AtTag.class);
        tagClassMap.put(ExistsTag.name, ExistsTag.class);
        tagClassMap.put(ForEachTag.name, ForEachTag.class);
        tagClassMap.put(LenTag.name, LenTag.class);
        tagClassMap.put(OneOfTag.name, OneOfTag.class);
    }

    @SuppressWarnings("unchecked")
    public static Map<String, TagBase> getTagMap(TagResolver tagResolver) throws TemplateEngineException {
        Map<String, TagBase> tagMap = new HashMap<>();
        for (Map.Entry<String, Class<?>> entry : tagClassMap.entrySet()) {
            try {
                Constructor<TagBase> ctor = (Constructor<TagBase>) entry.getValue().getConstructor(
                        TagResolver.class);
                TagBase object = ctor.newInstance(tagResolver);
                tagMap.put(entry.getKey(), object);
            } catch (NoSuchMethodException | InstantiationException |
                    IllegalAccessException | InvocationTargetException e) {
                throw new TemplateEngineException("Failed to add tag " + entry.getKey(), e);
            }
        }
        return tagMap;
    }

    public static void addTag(String tagName, Class<?> tag) {
        tagClassMap.put(tagName, tag);
    }

    public static String[] getTagNames() {
        return tagClassMap.keySet().toArray(new String[0]);
    }
}
